toolchain/check/eval.cpp

```diff
@@ -286,37 +286,174 @@ static auto PerformAggregateIndex(Context& context, SemIR::Inst inst)
   return context.constant_values().Get(elements[index_val.getZExtValue()]);
 }
 
+// Issues a diagnostic for a compile-time division by zero.
+static auto DiagnoseDivisionByZero(Context& context, SemIRLocation loc)
+    -> void {
+  CARBON_DIAGNOSTIC(CompileTimeDivisionByZero, Error, "Division by zero.");
+  context.emitter().Emit(loc, CompileTimeDivisionByZero);
+}
+
+// Performs a builtin unary integer -> integer operation.
+static auto PerformBuiltinUnaryIntOp(Context& context, SemIRLocation loc,
+                                     SemIR::BuiltinFunctionKind builtin_kind,
+                                     SemIR::InstId arg_id)
+    -> SemIR::ConstantId {
+  CARBON_CHECK(builtin_kind == SemIR::BuiltinFunctionKind::IntNegate)
+      << "Unexpected builtin kind";
+
+  auto op = context.insts().GetAs<SemIR::IntLiteral>(arg_id);
+  auto op_val = context.ints().Get(op.int_id);
+
+  if (op_val.isMinSignedValue()) {
+    CARBON_DIAGNOSTIC(CompileTimeIntegerNegateOverflow, Error,
+                      "Integer overflow in negation of {0}.", llvm::APSInt);
+    context.emitter().Emit(loc, CompileTimeIntegerNegateOverflow,
+                           llvm::APSInt(op_val, false));
+  }
+  op_val.negate();
+
+  auto result = context.ints().Add(op_val);
+  return MakeConstantResult(context, SemIR::IntLiteral{op.type_id, result},
+                            Phase::Template);
+}
+
+// Performs a builtin binary integer -> integer operation.
+static auto PerformBuiltinBinaryIntOp(Context& context, SemIRLocation loc,
+                                      SemIR::BuiltinFunctionKind builtin_kind,
+                                      SemIR::InstId lhs_id,
+                                      SemIR::InstId rhs_id)
+    -> SemIR::ConstantId {
+  auto lhs = context.insts().GetAs<SemIR::IntLiteral>(lhs_id);
+  auto rhs = context.insts().GetAs<SemIR::IntLiteral>(rhs_id);
+  auto lhs_val = context.ints().Get(lhs.int_id);
+  auto rhs_val = context.ints().Get(rhs.int_id);
+
+  bool overflow = false;
+  llvm::APInt result_val;
+  llvm::StringLiteral op_str = "<error>";
```

Comment:
This value should never be used, right? Should it be assigned at all?

Reply:
It should never be used, but `StringLiteral` doesn't have a default constructor.